### PR TITLE
fix(web): make jsonStringify circular-safe so event serialization never throws

### LINKED ISSUE #4132: jsonStringify throws on circular references (DOM nodes with React fibers); with capture_exceptions the throw is recaptured in a loop
# `jsonStringify` throws on circular references; with `capture_exceptions` the throw is recaptured in a loop

## Environment
- `posthog-js` 1.396.5 (also present on `main` / 1.399.2)
- React 17, Chrome
- `capture_exceptions: true`, autocapture enabled

## Symptom
A burst of ~100 identical `$exception` events in ~2 seconds, all `handled: true`, with value:

```
'TypeError' captured as exception with message:
'Converting circular structure to JSON
  --> starting at object with constructor 'HTMLAnchorElement'
  | property '__reactFiber$…' -> object with constructor 'FiberNode'
  --- property 'stateNode' closes the circle'
```

The frames resolve into the SDK's request/serialization path, not application code.

## Root cause
`jsonStringify` in `packages/browser/src/request.ts` — the serializer every request body passes through — only guards `BigInt`:

```ts
return JSON.stringify(data, (_, value) => (typeof value === 'bigint' ? value.toString() : value), space)
```

When a captured property holds a circular value, `JSON.stringify` throws `Converting circular structure to JSON`. The most common source on a React app is a DOM node: React attaches `__reactFiber$…` / `__reactProps$…` to elements, and a fiber's `stateNode` points back at the element, so any element reference that reaches event properties is circular.

## Why it loops
With `capture_exceptions` enabled, the `TypeError` thrown by the serializer is itself autocaptured as a new `$exception`. If the original event is retried, or the condition recurs across elements, the SDK repeatedly reports its own serialization failure — turning one unserializable property into a flood of self-referential exceptions.

## Minimal repro
```js
const el = document.querySelector('a')      // any React-rendered element
posthog.capture('x', { el })                // el is circular via its React fiber
// → TypeError: Converting circular structure to JSON, recaptured as $exception
```

## Fix
Make `jsonStringify` fall back to a circular-safe pass when the fast path throws, so serialization never throws. PR attached. The fast (non-circular) path is unchanged, so shared-but-acyclic references keep their real values; only a genuine back-reference becomes `"[Circular]"`.

Two deeper follow-ups worth considering, out of scope for the fix:
1. Exclude errors originating from the SDK's own capture/serialization path from `capture_exceptions`, to prevent self-referential loops in general.
2. Prevent raw DOM nodes from entering event properties in the first place.

### PR DESCRIPTION
**Pull request by mikenicholls88**
## Problem

`jsonStringify` in `packages/browser/src/request.ts` — the serializer every request body passes through — only guards `BigInt`. When a captured property holds a circular value, `JSON.stringify` throws `Converting circular structure to JSON`.

The common trigger on a React app is a DOM node: React attaches `__reactFiber$…` / `__reactProps$…` to elements, and a fiber's `stateNode` points back at the element, so any element reference that reaches event properties is circular. With `capture_exceptions` enabled, the `TypeError` the serializer throws is itself autocaptured as a new `$exception` — and when the condition recurs it turns into a burst of `handled` self-referential exceptions. We saw ~100 identical `$exception`s in a ~2 second window in production.

Fixes #4132.

## Changes

`jsonStringify` now falls back to a circular-safe replacer **only when the fast path throws**, so serialization always succeeds instead of throwing:

- The common (non-circular) path is byte-for-byte unchanged, so shared-but-acyclic references keep their real values.
- Only a genuine back-reference is replaced with `"[Circular]"`, matching the convention already used in `entrypoints/logs.ts`.
- `BigInt` handling is preserved (extracted to a shared replacer used by both paths).

No public API change; not a breaking change.

### Libraries affected

- [x] posthog-js (web)

## Checklist

- [x] Tests for new code
- [x] Accounted for the impact of any changes across different platforms
- [x] Accounted for backwards compatibility of any changes (no breaking changes!)
- [x] Took care not to unnecessarily increase the bundle size

### If releasing new changes

- [x] Ran `pnpm changeset` to generate a changeset file

## 🤖 Agent context

**Autonomy:** Human-driven (agent-assisted)

The bug was found while investigating an exception-tracking burst on a production React (17) / Next.js site running posthog-js: ~100 `handled` `$exception`s within ~2s, all `Converting circular structure to JSON` starting at an `HTMLAnchorElement` via its `__reactFiber$…` back-reference. Tracing the failing frame led to `jsonStringify` in `request.ts`, whose replacer guards `BigInt` but not circular references.

Approach considered and chosen: a whole-tree circular-safe replacer (WeakSet) on every call was rejected because it changes serialization of shared-but-acyclic references (a shared object would serialize as `"[Circular]"` on its second visit). Instead the fast path is left untouched and the circular-safe pass only runs after a throw, so normal payloads are unaffected and only genuinely-circular data is rewritten.

Authored with Claude Code (Claude). Directed and owned by @mikenicholls88.

### Test strategy

Added two cases to `request.test.ts` (body encoding) and ran the suite for the changed file:

```
pnpm --filter posthog-js exec jest src/__tests__/request.test.ts
→ Test Suites: 1 passed; Tests: 57 passed (55 existing + 2 new)
```

- **plain circular object** — asserts the exact `{"foo":"bar","self":"[Circular]"}` output (no throw).
- **circular DOM node** — a node whose `__reactFiber.stateNode` points back at the node (the shape that triggered this in production); asserts serialization does not throw and the request still sends.

Edge cases considered: `BigInt` handling is unchanged (existing test still passes); shared-but-acyclic references are unaffected because the fallback only runs after the fast path throws — verified separately that `{ a: shared, b: shared }` keeps both real values rather than emitting `[Circular]`.

Not done: I did not run the full monorepo suite or smoke-test in a browser example app. The change is confined to `jsonStringify` and covered by the unit tests above.

### CODE REVIEW

**Comment by mikenicholls88** [comment]
Extra evidence from the production instance where this surfaced (posthog-js 1.396.5, React 17 / Next.js 12):

- **108 `$exception` events in a ~2 second window**, all `mechanism.handled: true`, from a single browser session.
- Identical value on every one:
  ```
  'TypeError' captured as exception with message: 'Converting circular structure to JSON
    --> starting at object with constructor 'HTMLAnchorElement'
    | property '__reactFiber$…' -> object with constructor 'FiberNode'
    --- property 'stateNode' closes the circle'
  ```
- The failing frames resolved into the SDK's request/serialization chunk, not application code.

The shape (identical error, one session, sub-second cadence) is the `capture_exceptions`-recaptures-its-own-throw loop — one unserializable property becomes a flood. A `before_send` hook dropping `/Converting circular structure to JSON/` stops it as a consumer-side workaround, but the fix in this PR removes the throw at the source.

File: packages/browser/src/request.ts [hunk 1]
```diff
@@ -89,13 +89,40 @@ export const extendURLParams = (url: string, params: Record<string, any>, replac
...
+    } catch {
+        // A circular reference (e.g. a DOM node that retains a React fiber, or any other
+        // self-referential value that ends up in a captured property) makes JSON.stringify throw
+        // "Converting circular structure to JSON". That throw previously escaped the send path, and
+        // with exception autocapture enabled it would be recaptured as a new $exception — sometimes
+        // in a tight loop. Retry with a circular-safe replacer so serialization always succeeds
+        // rather than throwing. Only the offending back-reference is replaced with "[Circular]"; the
+        // common (non-circular) path above is unaffected, so shared-but-acyclic references keep their
+        // real values.
+        const seen = new WeakSet<object>()
```
Comment (turnipdabeets) [inline packages/browser/src/request.ts]: Could we make the catch just `return safeJsonStringify(data)` from `@posthog/core`? It was added in #4027 to fix this same bug in posthog-node and already guards the Node/RN capture paths, so this keeps one circular-serialization convention across SDKs (it also handles BigInt, and flattens Errors).

File: packages/browser/src/__tests__/request.test.ts [hunk 2]
```diff
@@ -713,6 +713,36 @@ describe('request', () => {
...
+                expect(() =>
+                    request(
+                        createRequest({
+                            url: 'https://any.posthog-instance.com/',
+                            method: 'POST',
+                            data: { $el: el },
+                        })
+                    )
+                ).not.toThrow()
+                expect(mockedXHR.send).toHaveBeenCalledTimes(1)
```
Comment (turnipdabeets) [inline packages/browser/src/__tests__/request.test.ts:744]: [suggestion] Worth a third case with a shared-but-acyclic sibling inside a circular payload (`const shared = {n: 1}; const data: any = {a: shared, b: shared}; data.self = data`) asserting `a` and `b` both keep their real values — that pins down the "only genuine back-references" behavior the PR description promises.

**Review by turnipdabeets — COMMENTED** [review]
Thanks for your contribution! A few comments below plus we need to make lint happy on CI and all of your commits need to be signed to merge. Otherwise LGTM

**Comment by mikenicholls88** [comment]
Thanks for the review — all addressed:

- **`safeJsonStringify`**: the `catch` now returns `safeJsonStringify(data)` from `@posthog/core` and I dropped the local `WeakSet`. Good call — its ancestor-chain approach handles the shared-but-acyclic case correctly, which the previous `WeakSet` version would have gotten wrong, and it keeps one circular-serialization convention across SDKs.
- **Third test**: added `keeps shared-but-acyclic references while replacing only true cycles` — `{ a: shared, b: shared }` with a `self` cycle, asserting `a`/`b` both keep their real values and only `self` becomes `"[Circular]"`.
- **Lint**: `eslint src/request.ts src/__tests__/request.test.ts` is clean.
- **Tests**: `jest src/__tests__/request.test.ts` → 58 passed.
- **Commits**: squashed to a single SSH-signed commit (registering the signing key on my account so it verifies).

**Comment by turnipdabeets** [comment]
@mikenicholls88 commits are signed but showing not verified, mind force pushing after registering on GitHub?

**Review by turnipdabeets — APPROVED** [review]
@mikenicholls88 thanks for your contribution! One request below and also could you
